model/qti/parser/IncludedElementIdsExtractor.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+/*
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoQtiItem\model\qti\parser;
+
+use oat\oatbox\service\ConfigurableService;
+use oat\tao\model\media\TaoMediaResolver;
+use oat\taoQtiItem\model\qti\Item;
+use tao_helpers_Uri;
+
+class IncludedElementIdsExtractor extends ConfigurableService
+{
+    public const OPTION_MEDIA_RESOLVER = 'mediaResolver';
+
+    /** @var string[] */
+    private $ids;
+
+    public function extract(Item $qtiItem): array
+    {
+        $this->ids = [];
+
+        $this->incrementIds($this->normalizeItemBody($qtiItem));
+
+        return array_unique($this->ids);
+    }
+
+    private function normalizeItemBody(Item $qtiItem): array
+    {
+        return json_decode(json_encode($qtiItem->toArray()['body']), true);
+    }
+
+    private function incrementIds(array $body): void
+    {
+        foreach ($body as $key => $value) {
+            $this->incrementIdsByKey($key, $value);
+        }
+    }
+
+    private function incrementIdsByKey(string $key, $value): void
+    {
+        if (!is_array($value)) {
+            return;
+        }
+
+        if ('elements' !== $key) {
+            $this->incrementIds($value);
+
+            return;
+        }
+
+        foreach ($value as $element) {
+            $this->incrementIdsByElement((array)$element);
+        }
+    }
+
+    private function incrementIdsByElement(array $element): void
+    {
+        if ($this->hasIncludedElement($element)) {
+            $this->ids[] = $this->formatElementId($element['attributes']['href']);
+        }
+
+        $this->incrementIds($element);
+    }
+
+    private function hasIncludedElement(array $element): bool
+    {
+        return ($element['qtiClass'] ?? '') === 'include' && !empty($element['attributes']['href']);
+    }
+
+    private function formatElementId(string $url): string
+    {
+        $id = $this->getMediaResolver()
+            ->resolve($url)
+            ->getMediaIdentifier();
+
+        return tao_helpers_Uri::decode($id);
+    }
+
+    private function getMediaResolver(): TaoMediaResolver
+    {
+        return $this->getOption(self::OPTION_MEDIA_RESOLVER) ?? new TaoMediaResolver();
```

Comment:
Where are u defining that this service is this configuration? u miss to configure at update and install

Reply:
I wanted to make it configurable for tests only, cause `TaoMediaResolver` does not extend ConfigurableService. 

What is the best option? 

- TaoMediaResolver extends ConfigurableService
- TaoMediaResolver at the Update script?

Reply:
> TaoMediaResolver extends ConfigurableService
Cannot because children are using construct args (vs configurableService $options)
> TaoMediaResolver at the Update script?
Why not, and at install. It will registered EventDispatcher with associated resolver config
> new TaoResolver()
It is not the best solution, but it requires minimal effort and do not impact actual behavior

Welcome in TAO dependencies injections

Reply:
Ok, I decided to use the third option, cause I do not want to make it configurable by user as option if it is not needed. I used DI pattern. Thanks!!